### PR TITLE
Prevent races creating web api session context

### LINKED ISSUE #23533: GRPC Errors handling in UI on initial login and access request applications
Expected behavior:

Errors like this would be caught the system would reattempt the page load w/o requiring user action.  The display of No Nodes Found is also incorrect.  If a page can't load due to a client error would expect an error shown of unable to load, not that there aren't any nodes.

There's also the ux of seeing a GRPC error as a output.  There are no additional messages to a user on what to do with this.  Most users would not know that we use GRPC and what this means.

Current behavior:

When users get this error they need to refresh or click to another tab. 

image

This happens whether a user is a local or SAML user. 

Bug details:
- Teleport version: 12.1.1

### PR DESCRIPTION
**Pull request by rosstimothy**
Closes #23533 and #20963.

There was a race to create the `web.SessionContext` for a session when multiple Proxies are behind a load balancer. Only the Proxy that processes the login will have a `web.SessionContext` created for the session. Any subsequent requests to the other Proxies in the pool would create one if the request was authenticated. However, multiple requests within short succession could cause a single Proxy to create multiple `web.SessionContext` for that session. When that happens the most recently created `web.SessionContext` gets saved and the previous `web.SessonContext` gets closed. Closing causes the `auth.Client` to be closed, which causes any active requests for that client to return with a `grpc: client connection is closing` error. This manifests in a single request from the web UI to fail and depending on the request, for a banner to be displayed with the error. Refreshing the page or navigating to another page would resolve the problem because the most recent `web.SessionContext` would be used with the still open `auth.Client`.

To prevent `web.Handler.AuthenticateRequest` from racing to create the `web.SessionContext` a `singleflight.Group` was added to the `web.sessionCache`. When multiple requests come in for the same session they now will all use the first `web.SessionContext` to be created instead of each creating their own.

### CODE REVIEW

File: lib/web/sessions.go [hunk 1]
```diff
@@ -793,17 +797,32 @@ func (s *sessionCache) ValidateTrustedCluster(ctx context.Context, validateReque
 	return s.proxyClient.ValidateTrustedCluster(ctx, validateRequest)
 }
 
-// validateSession validates the session given with user and session ID.
-// Returns a new or existing session context.
+// validateSession gets the SessionContext for the user and session ID. If one does
+// not exist, then a new one is created.
 func (s *sessionCache) validateSession(ctx context.Context, user, sessionID string) (*SessionContext, error) {
```
Comment (zmb3) [inline lib/web/sessions.go:802]: Should we rename this to something like `getOrCreateSession`?

File: lib/web/sessions.go [hunk 2]
```diff
@@ -793,17 +797,32 @@ func (s *sessionCache) ValidateTrustedCluster(ctx context.Context, validateReque
...
+	})
+
+	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return s.newSessionContext(ctx, user, sessionID)
+
+	sctx, ok := i.(*SessionContext)
+	if !ok {
+		return nil, trace.BadParameter("")
```
Comment (zmb3) [inline lib/web/sessions.go:821]:
```suggestion
		return nil, trace.BadParameter("expected SessionContext, got %T", i)
```

**Review by zmb3 — APPROVED** [review]

File: lib/web/sessions.go [hunk 3]
```diff
@@ -658,6 +659,9 @@ type sessionCache struct {
 	// sessions maps user/sessionID to an active web session value between renewals.
 	// This is the client-facing session handle
 	sessions map[string]*SessionContext
+	// sessionGroup ensures only a single SessionContext will exist for a
+	// user+session.
+	sessionGroup *singleflight.Group
```
Comment (espadolini) [inline lib/web/sessions.go:664]: Wouldn't this work?
```suggestion
	sessionGroup singleflight.Group
```

File: lib/web/sessions.go [hunk 4]
```diff
@@ -793,17 +797,32 @@ func (s *sessionCache) ValidateTrustedCluster(ctx context.Context, validateReque
...
-		return sessionCtx, nil
-	}
-	if !trace.IsNotFound(err) {
+// getOrCreateSession gets the SessionContext for the user and session ID. If one does
+// not exist, then a new one is created.
+func (s *sessionCache) getOrCreateSession(ctx context.Context, user, sessionID string) (*SessionContext, error) {
+	key := sessionKey(user, sessionID)
+
+	// Use sessionGroup to prevent multiple requests from racing to create a SessionContext.
+	i, err, _ := s.sessionGroup.Do(key, func() (interface{}, error) {
```
Comment (espadolini) [inline lib/web/sessions.go:806]: nit
```suggestion
	i, err, _ := s.sessionGroup.Do(key, func() (any, error) {
```

**Review by espadolini — APPROVED** [review]

**Pull request merged**